<a href="https://colab.research.google.com/github/Shubhankar9934/Shubhankar9934/blob/main/Word_Embedding%20Layers%20For%20Deep%20Learning%20With%20Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 **Word Embeddding Techniques Using Embedding Layer in Keras**

In [4]:
# Libraries used Tensorflow >2.0 and Keras
!pip install Tensorflow
!pip install keras

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [7]:
## Tensorflow >2.0
from tensorflow.keras.preprocessing.text import one_hot

In [8]:
## Sentences 
sent = ['the glass of milk',
        'the glass of juice',
        'the cup of tea',
        'I am a good boy',
        'understand the meaning of words',
        'your videos are good',]

In [22]:
# Sentences
sent

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'understand the meaning of words',
 'your videos are good']

In [21]:
## Vocabulary Size
voc_size = 10000

**One Hot Representation**


In [11]:

## One Hot Representation Gives Index from the Dictionary.
onehot_repr=[one_hot(words,voc_size)for words in sent] 
print(onehot_repr)

[[4994, 6784, 7466, 2988], [4994, 6784, 7466, 6748], [4994, 9423, 7466, 9259], [2387, 8677, 6157, 6415, 8912], [5994, 4994, 8853, 7466, 915], [841, 5354, 1455, 6415]]


**Word Embedding Represntation**

In [12]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential

In [13]:
import numpy as np

In [24]:

# One Hot Representation passing through Embedding Layer in Keras to form 
# Embedding Matrix.
sent_length=8 #( Here we are sending maximum length of sentences so that the 
#                 pad_sequences must have kept all the sentences having same word)
embedded_docs=pad_sequences(onehot_repr,padding='pre',maxlen=sent_length)
# ( See in the output here is four words but adding padding param we convert into 
#  8 words sentences)
# Here we are using 'pre' that's why padding is adding previous to the word.
# if we are using 'post' then padding is adding after the words
print(embedded_docs)

[[   0    0    0    0 4994 6784 7466 2988]
 [   0    0    0    0 4994 6784 7466 6748]
 [   0    0    0    0 4994 9423 7466 9259]
 [   0    0    0 2387 8677 6157 6415 8912]
 [   0    0    0 5994 4994 8853 7466  915]
 [   0    0    0    0  841 5354 1455 6415]]


In [15]:
# Must Have to assign the Dimension of the matrix because we are doing feature 
# Representation.
dim=10

In [16]:
model=Sequential()
model.add(Embedding(voc_size,10,input_length=sent_length)) # Here model is added embeding
model.compile('adam','mse')# (using adam optimizer and preformance matrix mean squared error (mse))

In [17]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 8, 10)             100000    
                                                                 
Total params: 100,000
Trainable params: 100,000
Non-trainable params: 0
_________________________________________________________________


In [18]:
print(model.predict(embedded_docs))

[[[-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801
    0.04002908 -0.02790256 -0.03956247  0.01538916 -0.01355677]
  [-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801
    0.04002908 -0.02790256 -0.03956247  0.01538916 -0.01355677]
  [-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801
    0.04002908 -0.02790256 -0.03956247  0.01538916 -0.01355677]
  [-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801
    0.04002908 -0.02790256 -0.03956247  0.01538916 -0.01355677]
  [-0.02727523 -0.04132672 -0.03071368 -0.00578097  0.0259596
   -0.01937931  0.02401114  0.009759   -0.00656857  0.00147909]
  [ 0.03407836  0.00523255  0.01055221  0.02348055  0.03702087
    0.02171155  0.02844366  0.03412673 -0.01270262 -0.01429207]
  [-0.04888978  0.02594057 -0.01645416  0.01257533 -0.04057191
    0.02984525  0.04364684 -0.03007834 -0.01696737 -0.03457423]
  [ 0.01465262  0.0106729  -0.02858303  0.04869552  0.04621622
   -0.03861774  0.03016387  0.00544106  0.0162786

In [19]:
embedded_docs[0]

array([   0,    0,    0,    0, 4994, 6784, 7466, 2988], dtype=int32)

In [20]:
# Let's See Here Every Words in embedded_docs[0] is converted into 10 vectors.
print(model.predict(embedded_docs)[0])

[[-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801  0.04002908
  -0.02790256 -0.03956247  0.01538916 -0.01355677]
 [-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801  0.04002908
  -0.02790256 -0.03956247  0.01538916 -0.01355677]
 [-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801  0.04002908
  -0.02790256 -0.03956247  0.01538916 -0.01355677]
 [-0.03136565 -0.01105386  0.01131023  0.00725819  0.00702801  0.04002908
  -0.02790256 -0.03956247  0.01538916 -0.01355677]
 [-0.02727523 -0.04132672 -0.03071368 -0.00578097  0.0259596  -0.01937931
   0.02401114  0.009759   -0.00656857  0.00147909]
 [ 0.03407836  0.00523255  0.01055221  0.02348055  0.03702087  0.02171155
   0.02844366  0.03412673 -0.01270262 -0.01429207]
 [-0.04888978  0.02594057 -0.01645416  0.01257533 -0.04057191  0.02984525
   0.04364684 -0.03007834 -0.01696737 -0.03457423]
 [ 0.01465262  0.0106729  -0.02858303  0.04869552  0.04621622 -0.03861774
   0.03016387  0.00544106  0.01627864 -0.04637579]]

here we are observing that the first 4 value is same beacause starting value in embeded_docs[0] is same i.e 0.
array([   0,    0,    0,    0, 4994, 6784, 7466, 2988], dtype=int32). Afterwords all words is converted.

In [25]:
# https://machinelearningmastery.com/use-word-embedding-layers-deep-learning-keras/

# All intutions has been drawn from here.